Assignment3/Q1.py

```diff
@@ -1,128 +1,125 @@
+            i: an integer, first coordinate of an entry
+            j: an integer, second coordinate of an entry
+            rows: an integer, number of rows in a grid
+            columns: an integer, number of columns in a grid
+            visited: 2d boolean array, indicates which entries in a grid has already been visited (they are set to True)
 
         Returns:
-            a dictionary, adjacency list representation of the graph corresponding to the character grid given in
-            the problem. A vertex in a graph is an integer tuple which indicates the coordinates of an entry in
-            the grid.
-    """
-    # Using a dictionary data structure for the adjacency list representation of the graph.
-    graph = {}
-    for i in range(rows):
-        for j in range(columns):
-            # Add an entry (tuple of coordinates) for each cell in a grid.
-            graph[(i, j)] = []
-
-            # Create arrays of possible adjacent entries.
-            adjacent_horizontal = [i - 1, i, i + 1]
-            adjacent_vertical = [j - 1, j, j + 1]
+            a boolean, True if an entry is in a grid and has not been visited yet, False otherwise."""
+    return 0 <= i < rows and 0 <= j < columns and not visited[i][j]
 
-            # Populate each entry in a graph adjacency list representation.
-            for h in adjacent_horizontal:
-                for v in adjacent_vertical:
-                    # There are eight adjacent cells to each cell in a grid
-                    # Apart from edge cases, here we take care of edge cases.
-                    if 0 <= h < rows and 0 <= v < columns and not (h == i and v == j):
-                        graph[(i, j)] = graph[(i, j)] + [(h, v)]
 
-    # Return an adjacency list representation of a graph corresponding to a grid.
-    return graph
-
-
-def word_search(rows, columns, grid, dictionary):
-    """A method performs a word search on a character grid and returns a set of words (no duplicates) that can be formed
-       from grid's characters and are in a dictionary. Word search is case insensitive.
+def word_search(grid, dictionary):
+    """A method performs a word search on a character grid and returns a list of words (no duplicates) that can
+       be formed from grid's characters and are in a dictionary. Word search is case insensitive.
 
         Args:
-            rows: an integer, number of rows in a grid
-            columns: an integer, number of columns in a grid
             grid: a 2D array of characters, we are searching for words from dictionary in a grid
             dictionary: a Dictionary object, contains the words to search for in a grid
 
         Returns:
-            a set of words, complete list of words that can be formed from grid's characters and are in a dictionary,
-            there are no duplicated words in a set.
+            a list of words (all uppercase), complete list of words that can be formed from grid's characters and
```

Comment:
I changed it slightly and I'm using a set instead of the list the whole way thourgh.